# mbctools

We developed a python package called mbctools, designed to offer a cross-platform tool for processing amplicon data from various organisms in the context of metabarcoding studies. It can handle the most common tasks in metabarcoding pipelines like paired-end merging, primer trimming, quality filtering, zero-radius operational taxonomic units (ZOTU) denoising and filtering. This pipeline has the capability to process  multiple genetic markers simultaneously. mbctools is a menu-driven program that eliminates the need for expertise in command-line skills and ensures documentation of each analysis for reproducibility purposes. In our workflow, VSEARCH is utilized for processing fastq files derived from amplicon data. To adapt to the diversity of projects in metabarcoding, we facilitate the  reprocessing of  datasets with the possibility to adjust parameters. mbctools can also be launched in headless mode, making it suited for integration into pipelines running on High-Performance Computing (HPC) environments. mbctools is available at: [https://github.com/GuilhemSempere/mbctools](https://github.com/GuilhemSempere/mbctools).

Test data are available at https://doi.org/10.23708/W3TODJ. 